File: docs-examples/pubsub-python/src/test/python/micronaut/configuration/lettuce/docs/pubsub/BookEventRecorder.py
from typing import Annotated

from micronaut.configuration.lettuce.pubsub.annotation import MessageChannel, RedisListener
from micronaut.http import MediaType
from micronaut.http.annotation import Consumes
from micronaut.messaging.annotation import MessageBody

from .BookCreated import BookCreated


@RedisListener
class BookEventRecorder:
    """Records the messages published by the documented publisher and client, next to the documented listeners."""

    def __init__(self):
        self.created: list[str] = []
        self.plain: list[str] = []

    @MessageChannel("books.created")
    def created_event(self, event: Annotated[BookCreated, MessageBody], channel: Annotated[str, MessageChannel]) -> None:
        self.created.append(f"{channel}:{event.title}:{event.author}")

    @Consumes(MediaType.TEXT_PLAIN)
    @MessageChannel("books.plain-text")
    def plain_event(self, body: Annotated[str, MessageBody]) -> None:
        self.plain.append(body)

File: docs-examples/pubsub-kotlin/src/test/kotlin/io/micronaut/configuration/lettuce/docs/pubsub/BookEventRecorder.kt
package io.micronaut.configuration.lettuce.docs.pubsub

import io.micronaut.configuration.lettuce.pubsub.annotation.MessageChannel
import io.micronaut.configuration.lettuce.pubsub.annotation.RedisListener
import io.micronaut.http.MediaType
import io.micronaut.http.annotation.Consumes
import io.micronaut.messaging.annotation.MessageBody
import java.util.concurrent.BlockingQueue
import java.util.concurrent.LinkedBlockingQueue

/**
 * Records the messages published by the documented publisher and client, next to the documented listeners.
 */
@RedisListener
class BookEventRecorder {

    val created: BlockingQueue<String> = LinkedBlockingQueue()
    val plain: BlockingQueue<String> = LinkedBlockingQueue()

    @MessageChannel("books.created")
    fun created(@MessageBody event: BookCreated, @MessageChannel channel: String) {
        created.add("$channel:${event.title}:${event.author}")
    }

    @Consumes(MediaType.TEXT_PLAIN)
    @MessageChannel("books.plain-text")
    fun plain(@MessageBody body: String) {
        plain.add(body)
    }
}
